File: bongf/java/src/dongbinbook/ch18_graph_questions/Q41_TravelPlan.java
package dongbinbook.ch18_graph_questions;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class Q41_TravelPlan {
    private static int[] parent;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] input1 = br.readLine().split(" ");
        int n = Integer.parseInt(input1[0]);
        int m = Integer.parseInt(input1[1]);
        parent = new int[n+1];
        for (int i = 1; i <= n; i++) {
            parent[i] = i;
        }
        for (int i = 1; i < n+1; i++) {
            Integer[] data = Arrays.stream(br.readLine().split(" ")).map(Integer::parseInt).toArray(Integer[]::new);
            for (int j = 0; j < data.length; j++) {
                if(data[j] == 1) {
                    union(i, j+1);
                }
            }
        }
        Integer[] plans = Arrays.stream(br.readLine().split(" ")).map(Integer::parseInt).toArray(Integer[]::new);
        boolean canGo = true;
        for (int i = 0; i < m-1; i++) {
            if(findParent(plans[i]) != findParent(plans[i+1])) {
                canGo = false;
                break;
            }
        }
        if(canGo) {
            System.out.println("YES");
            return;
        }
        System.out.println("NO");
    }
    
    public static int findParent(int x) {
        if(parent[x] != x) {
            parent[x] = findParent(parent[x]);
        }
        return parent[x];
    }
    
    public static void union(int x, int y) {
        x = findParent(x);
        y = findParent(y);
        if(x < y) {
            parent[y] = x;
            return;
        }
        parent[x] = y;
    }
}

/*
5 4
0 1 0 1 1
1 0 1 1 0
0 1 0 0 0
1 1 0 0 0
1 0 0 0 0
2 3 4 3
 */

File: bongf/python/dongbinbook/ch18_graph_questions/q41_travel_plan.py
import sys
input = sys.stdin.readline
n, m = map(int, input().split())
parent = [0] * (n+1)
for i in range(1, n+1):
    parent[i] = i

def find_parent(x):
    if(parent[x] != x):
        parent[x] = find_parent(parent[x])
    return parent[x]

def union(x, y):
    x = find_parent(x)
    y = find_parent(y)
    if x < y : 
        parent[y] = x
    else:
        parent[x] = y

for i in range(1, n+1):
    data = list(map(int, input().split()))
    for j  in range(n):
        if data[j] == 1:
            union(i, j+1)

plan = list(map(int, input().split()))
can_go = True

print(parent)

for i in range(m-1):
    if find_parent(plan[i]) != find_parent(plan[i+1]):
        print(i, i+1)
        can_go = False

if can_go :
    print("YES")
else:
    print("NO")

'''
5 4
0 1 0 1 1
1 0 1 1 0
0 1 0 0 0
1 1 0 0 0
1 0 0 0 0 
2 3 4 3
'''
